File: 19_20 Runde2/Abbiegen/Quelltext/src/main/java/de/ncrypted/abbiegen/Dijkstra.java
package de.ncrypted.abbiegen;

import de.ncrypted.abbiegen.geometry.PointInt;
import javafx.util.Pair;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;

/**
 * @author ncrypted
 */
public class Dijkstra {

    public static Pair<Double, List<Integer>> dijkstra(List<PointInt> nodes, List<List<Pair<Integer, Double>>> graph, int start, int end) {
        List<Double> dist = new ArrayList<>(nodes.size());
        List<Integer> pred = new ArrayList<>(nodes.size());
        for (int i = 0; i < nodes.size(); i++) {
            dist.add(Integer.MAX_VALUE + 0.0);
            pred.add(-1);
        }
        dist.set(start, 0.0D);
        pred.set(start, -1);

        PriorityQueue<Pair<Double, Integer>> pq = new PriorityQueue<>((p1, p2) -> p1.getKey() > p2.getKey() ? 1 : -1);
        pq.add(new Pair<>(0.0D, start));
        while (!pq.isEmpty()) {
            int u = pq.peek().getValue();
            double uDist = pq.poll().getKey();
            if (dist.get(u) != uDist) {
                continue;
            }
            for (Pair<Integer, Double> edge : graph.get(u)) {
                int v = edge.getKey();
                double w = edge.getValue();
                if (uDist + w < dist.get(v)) {
                    dist.set(v, uDist + w);
                    pred.set(v, u);
                    pq.add(new Pair<>(uDist + w, v));
                }
            }
        }
        List<Integer> path = constructPath(pred, start, end);
        return new Pair<>(dist.get(end), path);
    }

    private static List<Integer> constructPath(List<Integer> pred, int start, int end) {
        List<Integer> path = new ArrayList<>();
        int cur = end;
        while (cur != start) {
            path.add(cur);
            cur = pred.get(cur);
        }
        path.add(start);
        Collections.reverse(path);
        return path;
    }
}

File: 19_20 Runde2/AbbiegenCPP/dijkstra.cpp
#include "dijkstra.h"
#include <queue>
#include <algorithm>

using std::priority_queue;
using std::make_pair;

typedef pair<double, int> diPair;

pair<double, vector<int>>
Dijkstra::dijkstra(vector<IntPoint> &nodes, vector<vector<pair<int, double>>> &graph, int start, int end) {
    int n = nodes.size();
    vector<double> dist(n, MAX_INT);
    vector<int> pred(n, -1);
    dist[start] = 0;

    priority_queue<diPair, vector<diPair>, std::greater<diPair>> pq;
    pq.push(make_pair(0, start));
    while (!pq.empty()) {
        int u = pq.top().second;
        double uDist = pq.top().first;
        if (dist[u] != uDist) {
            continue;
        }
        for (pair<int, double> edge : graph[u]) {
            int v = edge.first;
            int w = edge.second;
            if (uDist + w < dist[v]) {
                dist[v] = uDist + w;
                pred[v] = u;
                pq.push(make_pair(uDist + w, v));
            }
        }
    }
    vector<int> path = constructPath(pred, start, end);
    return make_pair(dist[end], path);
}

vector<int> Dijkstra::constructPath(vector<int> &pred, int start, int end) {
    vector<int> path;
    int cur = end;
    while (cur != start) {
        path.push_back(cur);
        cur = pred[cur];
    }
    path.push_back(start);
    std::reverse(path.begin(), path.end());
    return path;
}